dom5.ts

```diff
@@ -312,20 +312,34 @@ export function treeMap<U>(node: Node, mapfn: (node: Node) => U[]): U[] {
   return results;
 }
 
+export type childNodesRetriever = (node: Node) => Node[] | undefined;
+
+const defaultChildNodes: childNodesRetriever = node => node.childNodes;
+
+export const childNodesIncludeTemplate: childNodesRetriever = node => {
+  if (node.nodeName === 'template')
+    return treeAdapters.default.getTemplateContent(node).childNodes;
+  return node.childNodes;
+};
+
 /**
  * Walk the tree down from `node`, applying the `predicate` function.
  * Return the first node that matches the given predicate.
  *
  * @returns `null` if no node matches, parse5 node object if a node matches.
  */
-export function nodeWalk(node: Node, predicate: Predicate): Node|null {
+export function nodeWalk(
+    node: Node,
+    predicate: Predicate,
+    retrieveChildNodes: childNodesRetriever = defaultChildNodes): Node|null {
```

Comment:
I am not sure if I understand your question. The reason I chose to pass around a lambda is that 1. the user does not know, it is just a constant and 2. it prevents a lot of duplication in all methods that do traversal. Adding a lot of if-statements for each specific case felt redundant in my eyes. This lambda can be easily changed to change the behavior in all traversal methods.